### PR TITLE
Native implementation of aggregate function for FLOAT

### PR DESCRIPTION
**Pull request by maciejgrzybek**
It is to avoid confusion when coercions to DOUBLE are used and e.g. sum(float) = double.
E.g. when single value is given to sum(float) it would return double value which would not equal input when coercions are used.
Small commits added for review convenience, thus commit-by-commit approach is preferred.

### CODE REVIEW

**Comment by maciejgrzybek** [comment]
CC: @akshatnair, @arhimondr, @fiedukow

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java [hunk 1]
```diff
@@ -56,19 +56,61 @@ public static void updateCovarianceState(CovarianceState state, double x, double
...
 
+    public static double getCorrelation(CorrelationState state)
+    {
+        // Math comes from ISO9075-2:2011(E) 10.9 General Rules 7 c x
+        double dividend = state.getCount() * state.getSumXY() - state.getSumX() * state.getSumY();
+        dividend = dividend * dividend;
+        double divisor1 = state.getCount() * state.getSumXSquare() - state.getSumX() * state.getSumX();
+        double divisor2 = state.getCount() * state.getSumYSquare() - state.getSumY() * state.getSumY();
+
+        // divisor1 and divisor2 deliberately not checked for zero because the result can be Infty or NaN even if they are both not zero
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java:84]: is Infty a typo or deliberate abbreviation? just use Infinity full word if not a good reason

Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java:84]: of u just extracted it from older comment

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DoubleHistogramAggregation.java [hunk 2]
```diff
@@ -33,15 +33,15 @@
 import static com.facebook.presto.util.Failures.checkCondition;
 
 @AggregationFunction("numeric_histogram")
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DoubleHistogramAggregation.java:35]: Does the name Numeric actually imply decimal? or is it just any number format?

File: presto-main/src/test/java/com/facebook/presto/operator/TestFloatAverageAggregation.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        MetadataManager metadata = MetadataManager.createTestMetadataManager();
+        avgFunction = metadata.getFunctionRegistry().getAggregateFunctionImplementation(
+                new Signature("avg", FunctionKind.AGGREGATE, parseTypeSignature(StandardTypes.FLOAT), parseTypeSignature(StandardTypes.FLOAT)));
+    }
+
+    @Test
+    public void averageOfNullIsNull()
+    {
+        assertAggregation(avgFunction,
+                1.0,
```
Comment (akshatnair) [inline presto-main/src/test/java/com/facebook/presto/operator/TestFloatAverageAggregation.java:55]: may be extract your confidence value as a constant?

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestFloatSumAggregation.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    public Block[] getSequenceBlocks(int start, int length)
+    {
+        BlockBuilder blockBuilder = FLOAT.createBlockBuilder(new BlockBuilderStatus(), length);
+        for (int i = start; i < start + length; i++) {
+            FLOAT.writeLong(blockBuilder, floatToRawIntBits((float) i));
+        }
+        return new Block[] {blockBuilder.build()};
+    }
+
+    @Override
```
Comment (akshatnair) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestFloatSumAggregation.java:40]: Is there already a test of sum() using null cast as Float?

**Comment by akshatnair** [comment]
LGTM mod comments

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java (hunk 1 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java:84]: I just extracted this code from other place. It's a copy & paste.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/DoubleHistogramAggregation.java (hunk 2 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/DoubleHistogramAggregation.java:35]: This name was given already before. It worked on DOUBLE, so any type coercible to DOUBLE would benefit from this function.

File: presto-main/src/test/java/com/facebook/presto/operator/TestFloatAverageAggregation.java (hunk 3 above)
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/TestFloatAverageAggregation.java:55]: Other tests don't do this. I wanted to match code style.

Comment (akshatnair) [inline presto-main/src/test/java/com/facebook/presto/operator/TestFloatAverageAggregation.java:55]: I understand but if somewhere we have written a code that is less desirable so should we continue that or at least make sure that in newer code we do not repeat it? I will leave it to your judgement.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestFloatSumAggregation.java (hunk 4 above)
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestFloatSumAggregation.java:40]: It's tested in `AbstractTestAggregationFunction`. Have a look on e.g. `testMixedNullAndNonNullPositions` which generates null and non-null blocks alternating.

File: presto-main/src/test/java/com/facebook/presto/operator/TestFloatAverageAggregation.java (hunk 3 above)
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/TestFloatAverageAggregation.java:55]: I agree with you but in this case I claim confidence is not meant to change (by definition of these particular tests), so I'll leave it as is.

**Comment by arhimondr** [comment]
lgtm

**Comment by maciejgrzybek** [comment]
Thanks, merged to `sprint-34`.
Starting PR against `prestodb/master`.

**Pull request closed without merging**